src/Data/Bifunctor/ShowRead.hs

```diff
@@ -0,0 +1,123 @@
+  ( ShowRead (..)
+  , ShowRead1 (..)
+  , ShowRead2 (..)
+  , liftReadPrecWhatever
+  , liftShowsPrecWhatever
+  ) where
+import qualified Text.ParserCombinators.ReadPrec as TPR
+import qualified Text.Read.Lex as TRL
+import qualified Text.Read as TR
+import Text.Read (ReadPrec, Read (..), readListPrecDefault)
+import GHC.Generics
+import Data.Kind
+import Data.Functor.Classes
+
+newtype ShowRead a = ShowRead a
+newtype ShowRead1 f a = ShowRead1 (f a)
+newtype ShowRead2 f a b = ShowRead2 (f a b)
+
+instance (Wraps n d c s o, Read o) => Read (ShowRead n) where
+  readPrec = ShowRead <$> liftReadPrecWhatever readPrec
+  readListPrec = readListPrecDefault
+
+instance (Wraps n d c s o, Show o) => Show (ShowRead n) where
+  showsPrec d (ShowRead x) = liftShowsPrecWhatever showsPrec d x
+
+instance (Wraps1 n d c s o, Read1 o) => Read1 (ShowRead1 n) where
+  liftReadPrec rp rl = ShowRead1 <$> liftReadPrecWhatever @(n _) @d @c @s @(o _) (liftReadPrec rp rl)
+
+  liftReadListPrec = liftReadListPrecDefault
+
+instance (Wraps1 n d c s o, Show1 o) => Show1 (ShowRead1 n) where
+  liftShowsPrec sp sl d (ShowRead1 x) = liftShowsPrecWhatever @(n _) @d @c @s @(o _) (liftShowsPrec sp sl) d x
+
+instance (Wraps2 n d c s o, Read2 o) => Read2 (ShowRead2 n) where
+  liftReadPrec2 rp1 rl1 rp2 rl2 = ShowRead2 <$> liftReadPrecWhatever @(n _ _) @d @c @s @(o _ _) (liftReadPrec2 rp1 rl1 rp2 rl2)
+
+  liftReadListPrec2 = liftReadListPrec2Default
+
+instance (Wraps2 n d c s o, Show2 o) => Show2 (ShowRead2 n) where
+  liftShowsPrec2 sp1 sl1 sp2 sl2 d (ShowRead2 x) =
+    liftShowsPrecWhatever @(n _ _) @d @c @s @(o _ _) (liftShowsPrec2 sp1 sl1 sp2 sl2) d x
+
+type WrapsF n d c s o =
+  ( Generic n
+  , Rep n ~ D1 d (C1 c (S1 s (Rec0 o)))
+  , Constructor c
+  , Selector s )
+
+class WrapsF n d c s o => Wraps n d c s o
+instance WrapsF n d c s o => Wraps n d c s o
+
+type family Any where
+
+class (forall a. Wraps (n a) d c s (o a), Wraps (n Any) d c s (o Any)) => Wraps1 n d c s o
+instance (forall a. Wraps (n a) d c s (o a)) => Wraps1 n d c s o
+
+class (forall a b. Wraps (n a b) d c s (o a b), Wraps (n Any Any) d c s (o Any Any)) => Wraps2 n d c s o
+instance (forall a b. Wraps (n a b) d c s (o a b)) => Wraps2 n d c s o
+
+data Prox (c :: Meta) (f :: Type -> Type) a = Prox
+
+-- | Given a way to read the underlying type of a newtype
+-- or similar, produce a way to read the newtype itself
+-- using either record syntax or plain syntax.
+liftReadPrecWhatever
+  :: forall n d c s o.
+     ( Generic n
+     , Wraps n d c s o
+     , Constructor c
+     , Selector s)
+  => ReadPrec o -> ReadPrec n
+liftReadPrecWhatever read_p =
+    TR.parens $ do
+      expectP (TRL.Ident $ conName (Prox @c))
+      (TPR.prec 11 $ do
+         expectP (TRL.Punc "{")
+         expectP (TRL.Ident $ selName (Prox @s))
+         expectP (TRL.Punc "=")
+         p <- read_p
+         expectP (TRL.Punc "}")
+         pure (to (M1 (M1 (M1 (K1 p))))))
+        TR.+++
+          (TPR.prec 10 $ do
```

Comment:
Instead of hard-coding `10`, you could alternatively use [`appPrec`](https://hackage.haskell.org/package/base-4.15.0.0/docs/GHC-Show.html#v:appPrec) from `GHC.Show`.